Include bond forces in `compute stress/cartesian`

Compizfox opened the pull request:
**Summary**

This PR modifies `compute stress/cartesian` to include bond forces.

**Related Issue(s)**

This was discussed at the end of https://github.com/lammps/lammps/issues/3765#issuecomment-1534959286 and in [this](https://matsci.org/t/extending-stress-cartesian-to-include-bond-forces-how-to-grow-num-bond-bond-atom-arrays-to-include-ghost-atoms/48736) MatSci forum thread.

**Author(s)**

Lars Veldscholte, University of Twente, l.b.veldscholte@utwente.nl

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

Below the loop over pairs in the neighbor list, I added a loop over bonds (`neighbor->bondlist`). Similar to the above loop for pair forces, I skip the iteration if one of both atoms are not in the current group, and for ghost atoms I make sure that the computation is only done on the appropriate processor when `newton_bond == 0`. Barring that, I compute the bond force using `force->bond->single()` and call `compute_pressure()` using that force.

I also refactored the `compute_array()` and `compute_pressure()` methods somewhat, mostly by removing unnecessary copies of variables (especially if they're only used once) and by declaring the variables locally (instead of at the top of the method) so that they are properly scoped.

I checked the correctness using a system of polymers. The averaged pressure profiles correspond to the global pressure as output by `compute pressure`.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] One or more example input decks are included

Compizfox committed:
Include bond interactions in force

Compizfox committed:
Refactor compute_array() and compute_pressure():

Remove unnecessary copies of variables, declare variables locally so they are properly scoped

akohlmey commented on src/EXTRA-COMPUTE/compute_stress_cartesian.cpp:
```diff
@@ -229,35 +230,16 @@ void ComputeStressCartesian::init_list(int /* id */, NeighList *ptr)
-
   double **x = atom->x;
   double **v = atom->v;
-  double *mass = atom->mass;
```
Please keep these cached values from Atom class members and other pointers and values. It makes for an easier read of the code, keeps it consistent with other parts of LAMMPS, and has a (small) performance benefit since it reduces the need to chase pointers in every loop iteration.

Compizfox replied:
I sincerely doubt it makes any difference for performance (with modern compilers, but I'll be happy to be proved wrong ;)), but I'll revert the copied pointers for consistency with the rest of LAMMPS if so desired. :)

akohlmey reviewed (changes requested):
For backward compatibility, it would be preferred to have optional keywords (pair, bond, ...) that turn on individual contributions, similar to how it is done in compute pressure. Only the default setting would be "pair" to maintain backward compatibility. Please see the discussion of a very similar issue in PR #3792.
@athomps we should make this consistent across both pull requests.

athomps commented:
@Compizfox I think making bonded contributions optional is a good idea, unless it makes no sense to use fix stress/cartesian without bonded contributions. Is there a valid use case of this method for bonded systems in which only non-bonded force contributions are tallied by fix stress/cartesian?

Compizfox commented:
> @Compizfox I think making bonded contributions optional is a good idea, unless it makes no sense to use fix stress/cartesian without bonded contributions. Is there a valid use case of this method for bonded systems in which only non-bonded force contributions are tallied by fix stress/cartesian?

Hmm, I'm not sure if **not** taking into account bond forces for bonded systems ever makes sense. As far as I know, it simply results in a wrong force calculation, but maybe there are some edge cases that I don't know about where it does make sense. I can't think of any, though...

Maybe make it optional, but opt-out? [`compute stress/atom` also does this](https://docs.lammps.org/compute_stress_atom.html):

> If no extra keywords are listed, the kinetic contribution all of the virial contribution terms are included in the per-atom stress tensor. If any extra keywords are listed, only those terms are summed to compute the tensor.

Compizfox committed:
Revert removal of copies of pointers

athomps commented:
Yes, `stress/atom` is a good precedent to follow.  Can you also update doc page to mention that this compute only handles pair and bond contributions, so if contributions from other energy, such as kspace, angles etc. are used in the global virial calculation, these will not be accounted for by `stress/cartesian`

athomps reviewed (changes requested):
See comment.

Compizfox committed:
Fix shifted coordinates:

Add `boxlo` to the bin centers

Compizfox committed:
Add optional keywords to arguments for ke/pair/bond forces

Compizfox commented:
> Yes, `stress/atom` is a good precedent to follow. Can you also update doc page to mention that this compute only handles pair and bond contributions, so if contributions from other energy, such as kspace, angles etc. are used in the global virial calculation, these will not be accounted for by `stress/cartesian`

OK, working on it.

Would it make sense to split out the documentation for `compute stress/cartesian` from that of `compute stress/cylinder` and `compute stress/spherical`, since some of the stuff now only applies to the first and not to the other two? Otherwise, the combined documentation page might get a bit confusing maybe.

athomps commented:
> Would it make sense to split out the documentation for `compute stress/cartesian` from that of `compute stress/cylinder` and `compute stress/spherical`, since some of the stuff now only applies to the first and not to the other two? Otherwise, the combined documentation page might get a bit confusing maybe.

Ah, I forgot about those. Instead of breaking out the documents, it might not be much harder to add support for bonds to  `compute stress/cylinder` and `compute stress/spherical`

Compizfox commented:
> Ah, I forgot about those. Instead of breaking out the documents, it might not be much harder to add support for bonds to `compute stress/cylinder` and `compute stress/spherical`

I've looked into it, but `stress/cylinder` and `stress/spherical` don't exactly follow exactly the same structure as `stress/cartesian` (if they did, it would make sense for them to share significant code) and look considerably more complicated. Frankly, I do not immediately see how to also quickly add support for bonds for those two.

Compizfox committed:
Update documentation for compute stress/cartesian,

and split the doc page compute_stress_profile into compute_stress_cartesian and compute_stress_curvilinear

Compizfox commented:
@athomps @akohlmey is the PR alright like this?

akohlmey commented on doc/src/compute_stress_cartesian.rst:
```diff
@@ -0,0 +1,118 @@
+:math:`P^v`. The sum gives the total stress tensor :math:`P = P^k+P^v`.
+This compute obeys momentum balance through fluid interfaces. They use the
+Irving--Kirkwood contour, which is the straight line between particle pairs.
+
```
```suggestion

.. versionadded:: TBD

   Added support for bond styles

```

akohlmey commented:
> @athomps @akohlmey is the PR alright like this?

Not entirely. You need to correct the links in the manual that link to the compute_stress_profile.rst file that now link to compute_stress_curvilinear.rst. That would be at the very least in `compute.rst`, and `Commands_compute.rst`.
In the "Related commands" section you add mutual references to the compute styles in the two files. I've also added a suggestion for adding the `.. versionadded` tag to indicate the (future) release date of this modification, so people with older LAMMPS versions that use the online manual can see why their version may be missing this feature. TBD will be replaced in the last step before a release when version.h is updated.

Compizfox committed:
Update doc/src/compute_stress_cartesian.rst

Co-authored-by: Axel Kohlmeyer <akohlmey@gmail.com>

Compizfox committed:
Fix doc links

Compizfox commented:
> Not entirely. You need to correct the links in the manual that link to the compute_stress_profile.rst file that now link to compute_stress_curvilinear.rst. That would be at the very least in `compute.rst`, and `Commands_compute.rst`. In the "Related commands" section you add mutual references to the compute styles in the two files. I've also added a suggestion for adding the `.. versionadded` tag to indicate the (future) release date of this modification, so people with older LAMMPS versions that use the online manual can see why their version may be missing this feature. TBD will be replaced in the last step before a release when version.h is updated.

Ah, I missed that. Done.

Compizfox committed:
Merge branch 'develop' into develop

akohlmey commented:
> Ah, I missed that. Done.

From here on the final word is with @athomps. From the formal/technical point of view this is ready.

athomps reviewed (approved):
I approve.
